tests/test_assignment.py

```diff
@@ -76,3 +95,316 @@ def _prepare_submission(self):
 
         self.assertEqual(total_score, 0)
         self.assertEqual(max_score, 1)
+
+    def test_hard_fail(self):
+        # question, question, submission, score, graded_question, graded_question.
+        test_cases = [
+            # Incorrect submission, (fail, fail).
+            (
+                TestAssignment.QuestionBase(1), TestAssignment.QuestionBase(1),
+                False, 0,
```

Comment:
Using "submission" like you do here seems a little confusing and not very flexible (you can't have your two questions pass and soft fail).
Maybe instead, just make questions that always fail and pass (e.g. QuestionAlwaysPass, QuestionAlwaysFail).
Then you don't have to worry about prepping the submissions.

You also don't really need cases where you have a hard fail question that does not fail since we are not testing the code for making a test question (i.e. we are not testing line 26 in this file, which would control that behavior).